### ISSUE TITLE
Dnscap minor output issues

### ISSUE THREAD

**Issue by SuperAmbitiousItGuy** [DNS-OARC/dnscap#12]
I have found three minor issues concerning dnscap's output behavior:

First issuse:
Concerning option -s ir:

If only one of the two parameters (i or r) is used and the dns packets are sent "slowly" (e.g. 0.2+ seconds between each packet, achievable for example by executing several single queries using "dig" commands) **dnscap will not produce any output**. If there is little to no time between the packets, dnscap will produce output normally, although sometimes the first packet gets discarded. All of this of course only applies while using the -i option.

Second issue:
If dnscap is used to listen on a network interface (e.g. dnscap -i lo -g) **the very last packet will not be output**, which also means that in order to get any output at all, at least two packets have to be sent, since if only one packet is sent the first packet is also the last one. This is probably buffer-related.

Third issue:
Sending single dns queries (e.g. via dig, 0.2+ sec interval between the queries) only causes the dns **answer** packets to be output, instead of both the queries and the answers.

All three of these mentioned issues are fixed by using the **pcap_loop instead of the pcap_dispatch** function in dnscap.c. Is there any reason pcap_dispatch is used over pcap_loop?

Christian

**Comment by jelu** [comment]
All issues have been verified but currently quite busy with another thing so in a week or so I can start looking at this. The packet processing code is probably very old and needs updating.

**Comment by jelu** [comment]
And thanks for reporting this!

**Comment by wessels** [comment]
>  Is there any reason pcap_dispatch is used over pcap_loop?

No reason that I can think of or find.  Looking at the history, looks like pcap_dispatch() has been used since Paul Vixie's original code in 2007.

It seems to me that pcap_loop() could be made to work just as well.  Since dnscap supports reading from multiple interfaces, we can't use cnt=-1 with pcap_loop().

**Comment by jelu** [comment]
I have tracked this to be related to kernel packet filtering (BPF) and what kernel version is in use. @SuperAmbitiousItGuy can you specify what distro, kernel version and pcap library version you are using?

**Comment by jelu** [comment]
@SuperAmbitiousItGuy @wessels: `pcap_loop()` can not be used since it waits for packets to arrive and that will stall processing other interfaces until at least one packet arrives to the current monitored interface.

**Comment by SuperAmbitiousItGuy** [comment]
@jelu  My distro is Debian 8.5 (jessie), kernel version: 3.2.0.4-amd64 and my libpcap is 1.6.2-2 .
Thank you for looking into it, by the way!

**Comment by jelu** [comment]
Got a new libpcap engine in the works that will also support threads, just verified that it does not miss any packets on the same VM that did before. PR coming soon.

**Comment by jelu** [comment]
!["Good news, everyone!" -Farnsworth](http://1.bp.blogspot.com/-qNj_ZrdYoCw/UfrC9m1e8NI/AAAAAAAAYoI/zhLsOqZqkQk/s1600/futurama+professor+good+news+everyone.jpg)

Please try:
```
git clone https://github.com/jelu/dnscap.git
cd dnscap
git submodule update --init
sh autogen.sh
./configure
make
sudo src/dnscap [options...]
```

**Comment by aabdnn** [comment]
Exciting stuff! I'm busy testing our DNS server farm with heavy loads while dnscap is running, and I'll try to compare the single-threaded one to this one, and see if there's a difference.

**Comment by jelu** [comment]
Note that offline processing is currently missing from this branch.

@aabdnn Thanks for testing it, currently one thread is used per interface monitored so you should get twice the performance if monitoring one interface. Will add support for putting processing of packets in thread later and there is already interface for detecting packet loss because of lack of processing power, lot of stuff to add :)

**Comment by SuperAmbitiousItGuy** [comment]
@jelu I got your dnscap and executed the commands as you suggested but unfortunately the bug still occurs the same way as when i use the verisign/dnscap.

P.S. I'm also very excited to hear about all the new stuff that is to come!

**Comment by jelu** [comment]
@SuperAmbitiousItGuy Can you send the full console output from `git clone ...` to running it `sudo src/dnscap -g -d -d -d -d -d` and producing some output, to me jerry@dns-oarc.net

**Comment by jelu** [comment]
@SuperAmbitiousItGuy Question, how is it you are running kernel 3.2 when jessie came with 3.16 (https://www.debian.org/News/2015/20150426)

**Comment by jelu** [comment]
@aabdnn @SuperAmbitiousItGuy: Seems I forgot one thing in the instructions, need to checkout the correct branch so after `cd dnscap` please do `git checkout threads`, sorry for that!

**Comment by SuperAmbitiousItGuy** [comment]
@jelu It seems that the VM I am working on simply used the kernel of the underlying linux machine, which was 3.2. I have now downgraded it to 3.16.
I have sent an email with the said output.

EDIT:
I retried with `git checkout threads`
The bug seems to be resolved this way!
Thank you a lot for your efforts!

**Comment by jelu** [comment]
@SuperAmbitiousItGuy Great! Have a bit of work left before I can merge it into the code base, will close this issue when it is done.

**Comment by jelu** [comment]
Fixed as of 84923485e411ddc34fb7d49d087b4394996d1aa2
